File: client/lib/connection/impl/publickeypacket.dart
import 'package:client/connection/connection.dart';
import 'package:client/connection/impl/sessionkeypacket.dart';
import 'dart:typed_data';

import 'package:client/connection/packet.dart';
import 'package:client/crypto/aes.dart';
import 'package:client/crypto/rsa.dart';

class PublicKeyPacket extends Packet {
  PublicKey? _publicKey;

  PublicKeyPacket.write(this._publicKey);
  PublicKeyPacket.read() : _publicKey = null;

  @override
  int getId() {
    return 0;
  }

  @override
  Uint8List write() {
    return _publicKey!.serialize();
  }

  @override
  void read(Uint8List data) {
    _publicKey = PublicKey.deserialize(data);
  }

  @override
  void handle(Connection connection) {
    connection.setServerKey(_publicKey!);
    connection.setState(1);
    Key key = Key.generate();
    connection.setEncryptionCipher(AES(key));
    connection.send(SessionKeyPacket.write(key));
  }
}

File: server/Server/src/me/spencernold/server/connection/packet/impl/PublicKeyPacket.java
package me.spencernold.server.connection.packet.impl;

import me.spencernold.server.connection.Client;
import me.spencernold.server.connection.Client.State;
import me.spencernold.server.connection.packet.Id;
import me.spencernold.server.connection.packet.Packet;
import me.spencernold.server.encryption.PublicKey;
import me.spencernold.server.util.ByteBuffer;

@Id(id = 0)
public class PublicKeyPacket implements Packet {
	
	private PublicKey key;
	
	public PublicKeyPacket(PublicKey key) {
		this.key = key;
	}
	
	public PublicKeyPacket() {
	}

	public byte[] write() {
		return key.serialize();
	}

	public void read(ByteBuffer buf) throws Exception {
		key = PublicKey.deserialize(buf.toByteArray());
	}

	public void handle(Client client) {
		 client.data.setClientKey(key);
		 client.send(new PublicKeyPacket(client.server.data.getServerKeys().getPublicKey()));
		 client.state = State.HANDSHAKING;
	}
}
